sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -0,0 +1,305 @@
+        if not task.inputs.artifacts[input_name].producer_task:
+          artifact_type = _get_input_artifact_type(input_name,
+                                                   component_spec.inputs)
+
+          importer_task, importer_spec = importer_node.build_importer_spec(
+              task, input_name, artifact_type)
+          importer_tasks.append(importer_task)
+
+          task.inputs.artifacts[
+              input_name].producer_task = importer_task.task_info.name
+          task.inputs.artifacts[
+              input_name].output_artifact_key = importer_node.OUTPUT_KEY
+
+          deployment_config.executors[
+              importer_task.executor_label].importer.CopyFrom(importer_spec)
+
+    pipeline_spec.deployment_config.Pack(deployment_config)
+    pipeline_spec.tasks.extend(importer_tasks)
+
+    return pipeline_spec
+
+  def _create_pipeline(
+      self,
+      pipeline_func: Callable[..., Any],
+      pipeline_name: Optional[str] = None,
+      pipeline_description: Optional[str] = None,
+      params_list: List[dsl.PipelineParam] = None,
+  ) -> pipeline_spec_pb2.PipelineSpec:
+    """Internal implementation of create_pipeline.
+
+    Args:
+      pipeline_func: Pipeline function with @dsl.pipeline decorator.
+      pipeline_name: The name of the pipeline. Optional.
+      pipeline_description: An optional description of the pipeline.
+      params_list: A list of pipeline arguments.
+
+    Returns:
+      The IR representation (pipeline spec) of the pipeline.
+    """
+    params_list = params_list or []
+
+    # Create the arg list with no default values and call pipeline function.
+    # Assign type information to the PipelineParam
+    pipeline_meta = _python_op._extract_component_interface(pipeline_func)
+    pipeline_meta.name = pipeline_name or pipeline_meta.name
+    pipeline_meta.description = pipeline_description or pipeline_meta.description
+    pipeline_name = sanitize_k8s_name(pipeline_meta.name)
+
+    # Need to first clear the default value of dsl.PipelineParams. Otherwise, it
+    # will be resolved immediately in place when being to each component.
+    default_param_values = {}
+    for param in params_list:
+      default_param_values[param.name] = param.value
+      param.value = None
+
+    # Currently only allow specifying pipeline params at one place.
+    if params_list and pipeline_meta.inputs:
+      raise ValueError(
+          'Either specify pipeline params in the pipeline function, or in '
+          '"params_list", but not both.')
+
+    args_list = []
+    signature = inspect.signature(pipeline_func)
+    for arg_name in signature.parameters:
+      arg_type = None
+      for pipeline_input in pipeline_meta.inputs or []:
+        if arg_name == pipeline_input.name:
+          arg_type = pipeline_input.type
+          break
+      args_list.append(
+          dsl.PipelineParam(
+              sanitize_k8s_name(arg_name, True), param_type=arg_type))
```

Comment:
Again, not sure if we need this sanitization. how about drop now and add it when it's needed?

Reply:
@neuromage There is also a gotcha that even though the final pipeline structure might support arbitrary names, the string representations of the PipelineParam and other placeholders might result in problems.